app/controllers/onboarding_controller.rb

```diff
@@ -24,6 +27,31 @@ def create
 
   def success; end
 
+  # rubocop:disable Metrics/AbcSize
+  def telegram_auth
+    check_hash = telegram_auth_params[:hash]
+    auth_data = telegram_auth_params.slice(:auth_date, :first_name, :id, :last_name)
+    check_string = auth_data.to_unsafe_h.map { |k, v| "#{k}=#{v}" }.sort.join("\n")
+
+    secret_key = OpenSSL::Digest::SHA256.new.digest(ENV['TELEGRAM_BOT_API_KEY'])
+    hash = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha256'), secret_key, check_string)
+
+    raise ActionController::BadRequest if hash.casecmp(check_hash) != 0
+    raise ActionController::BadRequest if Time.now.to_i - telegram_auth_params[:auth_date].to_i > 86_400
+
+    @user = User.find_or_create_by(
+      telegram_id: telegram_auth_params[:id],
```

Comment:
When we have the telegram id through login, are we allowed to proactively send a message to the user?

Reply:
yes, we can ask for this permission with the `data-request-access="write">` part of the script.